WalletWasabi/Services/UpdateManager.cs

```diff
@@ -0,0 +1,276 @@
+using Newtonsoft.Json.Linq;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Net.Http;
+using System.Runtime.InteropServices;
+using System.Threading.Tasks;
+using WalletWasabi.Logging;
+using WalletWasabi.Microservices;
+using WalletWasabi.Models;
+using WalletWasabi.Tor.Http;
+
+namespace WalletWasabi.Services;
+
+public class UpdateManager : IDisposable
+{
+	private string InstallerPath { get; set; } = "";
+	private const byte MaxTries = 3;
+	private const string ReleaseURL = "https://api.github.com/repos/zkSNACKs/WalletWasabi/releases/latest";
+
+	public UpdateManager(string dataDir, bool downloadNewVersion, IHttpClient httpClient)
+	{
+		InstallerDir = Path.Combine(dataDir, "Installer");
+		HttpClient = httpClient;
+		DownloadNewVersion = downloadNewVersion;
+	}
+
+	private async void UpdateChecker_UpdateStatusChangedAsync(object? sender, UpdateStatus updateStatus)
+	{
+		var tries = 0;
+		bool updateAvailable = !updateStatus.ClientUpToDate || !updateStatus.BackendCompatible;
+		Version targetVersion = updateStatus.ClientVersion;
+		if (!updateAvailable)
+		{
+			Cleanup();
+			return;
+		}
+		if (DownloadNewVersion)
+		{
+			do
+			{
+				tries++;
+				try
+				{
+					(string installerPath, Version newVersion) = await GetInstallerAsync(targetVersion).ConfigureAwait(false);
+					InstallerPath = installerPath;
+					Logger.LogInfo($"Version {newVersion} downloaded successfuly.");
+					updateStatus.IsReadyToInstall = true;
+					updateStatus.ClientVersion = newVersion;
+					break;
+				}
+				catch (Exception ex)
+				{
+					Logger.LogError($"Geting new update failed with error.", ex);
+				}
+			} while (tries < MaxTries);
+		}
+
+		UpdateAvailableToGet?.Invoke(this, updateStatus);
+	}
+
+	/// <summary>
+	/// Get or download installer for the newest release.
+	/// </summary>
+	/// <param name="targetVersion">This does not contains the revision number, because backend always sends zero.</param>
+	private async Task<(string filePath, Version newVersion)> GetInstallerAsync(Version targetVersion)
+	{
+		(Version newVersion, string url, string fileName) = await GetLatestReleaseFromGithubAsync(targetVersion).ConfigureAwait(false);
+
+		var tmpFilePath = Path.Combine(InstallerDir, $"{fileName}.tmp");
+		var newFilePath = Path.Combine(InstallerDir, fileName);
+
+		var installerDownloaded = TryGetDownloadedInstaller(fileName);
+		if (!installerDownloaded)
+		{
+			EnsureToRemoveCorruptedFiles();
+
+			// This should also be done using Tor.
+			// TODO: https://github.com/zkSNACKs/WalletWasabi/issues/8800
+			Logger.LogInfo($"Trying to download new version: {newVersion}");
+			using HttpClient httpClient = new();
+
+			// Get file stream and copy it to downloads folder to access.
+			using var stream = await httpClient.GetStreamAsync(url).ConfigureAwait(false);
```

Comment:
This only creates the stream, which is instant, and Wasabi can shut down gracefully meanwhile the data is being downloaded, any leftover at the time of shutdown should be taken care of by GC/OS, worst thing that can happen I guess is an exception in the logs after exit.
On the other hand tho' if something needs a cancellation token it should be L90:
```cs
	await stream.CopyToAsync(file).ConfigureAwait(false);
```
Even tho' it shouldn't take much time and Wasabi can shutdown during copying as well, this can be a small improvement.

Reply:
If there is a place for the cancellation token, it should be used!